### PR TITLE
[RUNTIME] Refactor object python FFI to new protocol.

### PR DESCRIPTION
**Pull request by tqchen**
This is a pre-req to bring the Node system under object protocol.
Most of the code reflects the current code in the Node system.

- Use new instead of init so subclass can define their own constructors
- Allow register via name, besides type index
- Introduce necessary runtime C API functions
- Refactored Tensor and Datatype to directly use constructor.

### CODE REVIEW

**Comment by tqchen** [comment]
cc @icemelon9 @jroesch @wweic

**Comment by icemelon9** [comment]
will you shadow the python object by using filename object.py?

**Comment by tqchen** [comment]
I think we will not, given that object is not a keyword, as long as you do not do from tvm._ffi import object

File: python/tvm/_ffi/object.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        from ._cy2.core import _register_object
+except IMPORT_EXCEPT:
+    # pylint: disable=wrong-import-position
+    from ._ctypes.function import _set_class_object
+    from ._ctypes.object import ObjectBase as _ObjectBase
+    from ._ctypes.object import _register_object
+
+
+class Object(_ObjectBase):
+    """The VM Object used in Relay virtual machine."""
```
Comment (wweic) [inline python/tvm/_ffi/object.py]: Is this comment still accurate?

Comment (tqchen) [reply python/tvm/_ffi/object.py]: good point, will update it.

File: python/tvm/relay/backend/vmobj.py [hunk 2]
```diff
@@ -56,65 +61,34 @@ def asnumpy(self):
...
-        tensor = _vmobj.Tensor(_nd.array(arr, ctx))
-    elif isinstance(arr, _nd.NDArray):
-        tensor = _vmobj.Tensor(arr)
-    else:
-        raise RuntimeError("Unsupported type for tensor object.")
-    return tensor
+        return _vmobj.GetDatatypeNumberOfFields(self)
 
 
 def tuple_object(fields):
```
Comment (icemelon9) [inline python/tvm/relay/backend/vmobj.py]: change the doc to `ret : Datatype`

File: python/tvm/relay/backend/vmobj.py [hunk 3]
```diff
@@ -133,24 +107,3 @@ def tuple_object(fields):
     for f in fields:
         assert isinstance(f, Object)
     return _vmobj.Tuple(*fields)
```
Comment (icemelon9) [inline python/tvm/relay/backend/vmobj.py]: I feel that we can just use Datatype with tag=0 instead of using `_vmobj.Tuple`. it seems less confusing?

Comment (tqchen) [reply python/tvm/relay/backend/vmobj.py]: I think it would still be helpful to have an auxiliary function that constructs the tuple

Comment (icemelon9) [reply python/tvm/relay/backend/vmobj.py]: sure

**Comment by tqchen** [comment]
Thanks @wweic @icemelon9

**Pull request merged**